datafusion/common/src/scalar.rs

```diff
@@ -971,7 +1038,54 @@ impl ScalarValue {
 
                 Arc::new(StructArray::from(field_values))
             }
-            _ => {
+            DataType::Dictionary(key_type, value_type) => {
+                // create the values array
+                let value_scalars = scalars
+                    .into_iter()
+                    .map(|scalar| match scalar {
+                        ScalarValue::Dictionary(inner_key_type, scalar) => {
+                            if &inner_key_type == key_type {
+                                Ok(*scalar)
+                            } else{
+                                panic!("Expected inner key type of {} but found: {}, value was ({:?})", key_type, inner_key_type, scalar);
+                            }
+                        },
+                        _ => {
+                            Err(DataFusionError::Internal(format!(
+                                "Expected scalar of type {} but found: {} {:?}",
+                                value_type, scalar, scalar
+                            )))
+                        },
+                    })
+                    .collect::<Result<Vec<_>>>()?;
+
+                let values = Self::iter_to_array(value_scalars)?;
+                assert_eq!(values.data_type(), value_type.as_ref());
+
+                match key_type.as_ref() {
+                    DataType::Int8 => dict_from_values::<Int8Type>(&values)?,
+                    DataType::Int16 => dict_from_values::<Int16Type>(&values)?,
+                    DataType::Int32 => dict_from_values::<Int32Type>(&values)?,
+                    DataType::Int64 => dict_from_values::<Int64Type>(&values)?,
+                    DataType::UInt8 => dict_from_values::<UInt8Type>(&values)?,
+                    DataType::UInt16 => dict_from_values::<UInt16Type>(&values)?,
+                    DataType::UInt32 => dict_from_values::<UInt32Type>(&values)?,
+                    DataType::UInt64 => dict_from_values::<UInt64Type>(&values)?,
+                    _ => unreachable!("Invalid dictionary keys type: {:?}", key_type),
+                }
+            }
+            // explicitly enumerate unsupported types so newly added
```

Comment:
It took me a while to find that `ScalarValue::Dictionary` was missing here due to the default `_` match.